Add webcam streaming

huytrinhm opened the pull request:
Use vlc to stream from /dev/video0 to rtmp://localhost/live/webcam and use ffmpeg-webcam.service to save local chunks.

huytrinhm committed:
Add webcam streaming

huytrinhm committed:
test: webcam streaming

huytrinhm committed:
feat: remove webcam local recording

huytrinhm committed:
feat: reduce webcam streaming resolution to 720p

huytrinhm committed:
fix: ffmpeg record use port 101 from vlc

minhnhatnoe committed:
fix(dev_env): use xhci (usb 3.0)

minhnhatnoe commented:
Per internal communication, webcams are only available for access when the Virtual Machine's USB Controller is xHCI (USB 3.0) instead of the default OHCI (USB 2.0). I have set the development VM's USB Controller to xHCI. Great find!

huytrinhm committed:
fix: handle device disconnect

minhnhatnoe commented on image-toolkit/sbin/startup.sh:
```diff
@@ -67,7 +67,19 @@ webcam_stream_loop() {
+            }" &
+        CLVC_PID=$!
+        # Monitor the process and device
+        while [[ -e "/dev/video$DEVICE_NO" ]] && ps -p $CLVC_PID > /dev/null; do
```
You should monitor /dev/video with inotifywait (for example, `inotifywait -e delete_self /dev/video0` will exit when /dev/video0 is disconnected) and `wait -n <pidlist>` (wait for any in <pidlist> to finish).

Monitoring /dev/video0 by repeatedly polling could miss quick reconnection events (for example, a contestant accidentally touching the webcam's wire).

huytrinhm replied:
Thanks for your suggestion. I will update the code accordingly.

huytrinhm committed:
fix: use udevadm to monitor webcam events

huytrinhm commented:
We will use udevadm (from udev package already installed in Ubuntu 22.04) to listen for webcam-related events.

minhnhatnoe commented on image-toolkit/sbin/startup.sh:
```diff
@@ -53,6 +53,44 @@ vlc_restart_loop() {
+        udevadm monitor --udev -s video4linux | while read -r line; do
+            if [[ "$line" == *"remove"*"/video4linux/video$DEVICE_NO"* ]] ; then
+                echo "Device unplugged, proceed to kill cvlc"
+                kill $CLVC_PID 2>/dev/null || :
```
This logic is kinda complicated, and if VLC catch SIGTERM and fails to exit (like during startup), we'll be in a deadlock.

Let's make the udevadm process exit on device disconnect, then use `wait -n` to wait for the first process among the two to exit. If `udevadm` exits, we kill `vlc`. Exit udevadm also clears the event queue, so that consecutive connect-disconnect will not make us restart vlc more time than is necessary.

minhnhatnoe replied:
Also, for good measure, kill vlc with SIGKILL (`kill -9`) if it's not killed after a while (wait with something like `timeout wait $CVLC_PID`). Also, starting `udevadm` before `vlc` would be safer.

minhnhatnoe committed:
fix(startup.sh): vlc spelling

huytrinhm committed:
feat: include input audio in webcam streaming

huytrinhm committed:
fix: Set webcam keyframe interval to 15 fps

Using fixed keyframe interval significantly reduce stream latency and reduce *.ts fragment file size.

minhnhatnoe committed:
fix(startup.sh): move udevadm up before vlc to avoid missing video device remove event

minhnhatnoe committed:
feat(startup.sh): wait kills udevadm and cvlc only if terminated instead of status change

minhnhatnoe committed:
feat(startup.sh): udevadm specify video device

minhnhatnoe committed:
feat(startup.sh): prevent cpu hogging when video device not present and print debug

minhnhatnoe committed:
feat(startup.sh): kill -9 if cvlc fails to exit

minhnhatnoe committed:
fix(startup.sh): let kill terminate with exit code because we're not using exit on error

minhnhatnoe commented on image-toolkit/sbin/startup.sh:
```diff
@@ -53,6 +53,49 @@ vlc_restart_loop() {
 }
 
+webcam_stream_loop() {
+    VIDEO_DEVICE_NO=0
```
Due to https://github.com/obsproject/obs-studio/issues/3003 (webcam numbering may differ after reconnection), use devices in `/dev/v4l/by-id`

minhnhatnoe committed:
feat(startup.sh): pick webcams by id rather than /dev/video0

minhnhatnoe committed:
feat: process substitution instead of pipeline for udevadm

minhnhatnoe committed:
fix(startup.sh): correct timeout

minhnhatnoe committed:
feat: simplify process killing logic

minhnhatnoe committed:
feat: add sigkill to startup.sh's peer stop

huytrinhm committed:
Merge pull request #37 from VNOI-Admin/noe/webcam

Increases robustness of webcam_stream_loop

huytrinhm committed:
feat: Add audio device picker

huytrinhm committed:
feat: Add audio device utils for vnoiconf.sh

huytrinhm commented:
For audio streaming, we will have a configurable `AUDIO_DEVICE_NAME` in `config.sh`. When webcam streaming starts, it will search for a hardware audio capture device ([`/proc/asound/card*/pcm*c`](https://docs.kernel.org/sound/designs/procfile.html#card-specific-files)) and select the first one whose name matches `AUDIO_DEVICE_NAME`. Options for the system default device and the first hardware device are also available via `AUDIO_DEVICE_NAME=system` and `AUDIO_DEVICE_NAME=any`, respectively.

For ease of configuration, the `list_audio_devices` and `set_audio_device` commands are also made available in the `vnoiconf.sh` script. The VLC instance responsible for webcam streaming is automatically restarted after `set_audio_device` is used.

huytrinhm committed:
feat: allow glob and manual audio source config

minhnhatnoe committed:
feat: add comments to regexes in vnoiconf.sh

minhnhatnoe committed:
feat: pattern matching notes

minhnhatnoe committed:
feat: move default audio_source for clarity

hung3a8 committed:
chore: Cleanup github workflow jobs

hung3a8 committed:
fix: fix workflow image-toolkit config secret name

hung3a8 committed:
chore: drop root requirement when generating secrets

hung3a8 committed:
feat: Use workaround to gain extra space from github actions

hung3a8 committed:
fix: keep docker cache and image

hung3a8 committed:
feat: Remove IMAGE and ICPC folder in github actions mode

hung3a8 committed:
fix: Reorder github actions remove files

hung3a8 committed:
fix: use different tmp path for gh workflows scp actions

hung3a8 committed:
fix: Use homemade scp script

hung3a8 committed:
fix: please fix this scp madness

hung3a8 reviewed (approved):
LGTM
